pkg/reference/reference.go

```diff
@@ -95,11 +115,54 @@ func (rr ResolutionResponse) Validate() error {
 	return nil
 }
 
-// A Resolver selects and resolves references to managed resources.
-type Resolver interface {
-	// Resolve the supplied ResolutionRequest. The returned ResolutionResponse
-	// always contains valid values unless an error was returned.
-	Resolve(ctx context.Context, req ResolutionRequest) (ResolutionResponse, error)
+// A MultiResolutionRequest requests that several references to a particular
+// kind of managed resource be resolved.
+type MultiResolutionRequest struct {
+	CurrentValues []string
+	References    []v1alpha1.Reference
+	Selector      *v1alpha1.Selector
+	To            To
+	Extract       ExtractValueFn
+}
+
+// IsNoOp returns true if the supplied MultiResolutionRequest cannot or should
+// not be processed.
+func (rr MultiResolutionRequest) IsNoOp() bool {
+	// We don't resolve values that are already set; we effectively cache
+	// resolved values. The CR author can invalidate the cache and trigger a new
+	// resolution by explicitly clearing the resolved values. This is a little
+	// unintuitive for the APIMultiResolver but mimics the UX of the APIResolver
+	// and simplifies the overall mental model.
+	if len(rr.CurrentValues) > 0 {
+		return true
+	}
```

Comment:
Just making sure I am understanding what is happening here: if somehow some of the values were set for a multi-resolution reference (this should not be possible I don't think using `ResolveMultiple` below), then we would not try to set the remainder of them. The way to get out of this state would be to set clear the set values on the CR and let resolution try to set all of them again. Am I getting that right?

Reply:
Right.

I wondered whether this was the right approach, because I could imagine it being unintuitive to someone trying to add or update values by modifying the array of references. Some things I considered:

* I _think_ that in many cases the resolved values are immutable, even when they're an array. e.g. A Resource that accepts an array of subnet names often can't have have that array modified after the fact.
* Presuming we don't want to resolve all references on each reconcile it's tricky to know when we should cache them. Should we re-resolve when the array of references is longer than the array of values? Would this lead to an incorrect expectation that re-resolution would happen when a user edited an existing reference?

In the end it seemed conceptually simpler overall to maintain the same model as the single reference resolver; selector flows down to refs which flow down to values. Each is resolved once, and you need to "clear the cache" by removing the values, or the values and the refs, in order to trigger re-resolution of references.